src/ucp/core/ucp_ep.c

```diff
@@ -2667,3 +2669,74 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
         *lane_map &= ~UCS_BIT(lane);
     }
 }
+
+static void ucp_ep_req_purge(ucp_ep_h ucp_ep, ucp_request_t *req,
+                             ucs_status_t status, int recursive)
+{
+    if (req->id != UCP_REQUEST_ID_INVALID) {
+        ucp_request_id_release(req);
+    }
```

Comment:
is it possible some of the protocols cleanup flow would try to release the request it?

Reply:
nice catch, Yossi
currently, we don't have such protocols, but they could be implemented in the future
fixed this comment by moving checking/releasing of request ID at the end of the function, but need to implement some hack (save and remove RELEASE flag before calling request complete procedure)

Reply:
maybe "extract" the id from the request to a local variable?

Reply:
then protocols which expect that request ID is valid will fail `ucp_request_id_release()`